common/finalizer/finalizer.go

```diff
@@ -0,0 +1,170 @@
+	"go.temporal.io/server/internal/goro"
+)
+
+var (
+	FinalizerAlreadyDoneErr = errors.New("finalizer already finalized")
+	FinalizerUnknownIdErr   = errors.New("finalizer callback not found")
+	FinalizerDuplicateIdErr = errors.New("finalizer callback already registered")
+)
+
+type Finalizer struct {
+	logger    log.Logger
+	mu        sync.Mutex  // lock protects access to callbacks map
+	finalized atomic.Bool // indicating if finalizer has been executed
+	callbacks map[string]func(context.Context) error
+}
+
+func NewFinalizer(
+	logger log.Logger,
+) *Finalizer {
+	return &Finalizer{
+		logger:    logger,
+		callbacks: make(map[string]func(context.Context) error),
+	}
+}
+
+// Register adds a callback to the finalizer.
+// Returns an error if the ID is already registered, or when the finalizer already ran.
+func (f *Finalizer) Register(
+	id string,
+	callback func(context.Context) error,
+) error {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; ok {
+		return FinalizerDuplicateIdErr
+	}
+	f.callbacks[id] = callback
+	return nil
+}
+
+// Deregister removes a callback from the finalizer.
+// Returns an error if the ID is not found, or when the finalizer already ran.
+func (f *Finalizer) Deregister(
+	id string,
+) (err error) {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; !ok {
+		return FinalizerUnknownIdErr
+	}
+	delete(f.callbacks, id)
+	return nil
+}
+
+// Run executes all registered callback functions within the given timeout (zero timeout skips execution).
+// It can only be invoked once; calling it again has no effect.
+// Returns the number of completed callbacks.
+func (f *Finalizer) Run(
+	pool *goro.AdaptivePool,
+	timeout time.Duration,
+) int32 {
+	if !f.finalized.CompareAndSwap(false, true) {
+		f.logger.Warn("finalizer skipped: called more than once")
+		return 0
+	}
+
+	if timeout == 0 {
+		f.logger.Warn("finalizer skipped: zero timeout")
+		return 0
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	totalCount := int32(len(f.callbacks))
+	f.logger.Info("finalizer starting",
+		tag.NewInt32("items", totalCount),
+		tag.NewDurationTag("timeout", timeout))
+
+	completionChannel := make(chan struct{})
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	go func() {
+		for _, callback := range f.callbacks {
+			// NOTE: Once the pool is stopped due to a timeout, any remaining callbacks will not be invoked anymore.
+			pool.Do(func() {
+				defer func() { completionChannel <- struct{}{} }()
+				_ = callback(ctx)
+			})
+		}
+	}()
+
+	var completionCounter atomic.Int32
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-completionChannel:
+				completed := completionCounter.Add(1)
+				if completed == totalCount {
+					close(done)
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
```

Comment:
why this needs to be in a separate goroutine?

Reply:
I think you're right. This was using a WaitGroup before, but now that I've removed that, I can merge this select with the next select. I'll do that!